sunpy/map/mapbase.py

```diff
@@ -2811,7 +2818,33 @@ def plot(self, *, annotate=True, axes=None, title=True, autoalign=False,
         else:
             data = np.ma.array(np.asarray(self.data), mask=self.mask)
 
-        if autoalign == 'pcolormesh':
+        if autoalign == 'reproject':
+            target_wcs = copy.deepcopy(axes.wcs)
+
+            # Pull the reprojection extent from the plot limits
+            xlim, ylim = axes.get_xlim(), axes.get_ylim()
+
+            # If the plot limits are at the default values, use the pixel shape of the target WCS
+            if xlim == (0, 1) and ylim == (0, 1) and target_wcs.pixel_shape is not None:
+                xlim = (-0.5, target_wcs.pixel_shape[0] - 0.5)
+                ylim = (-0.5, target_wcs.pixel_shape[1] - 0.5)
+
+            # Expand the reprojection extent to whole pixels
+            left = int(np.floor(xlim[0] + 0.5))
+            bottom = int(np.floor(ylim[0] + 0.5))
+            right = int(np.ceil(xlim[1] - 0.5))
+            top = int(np.ceil(ylim[1] - 0.5))
+
+            # Modify the target WCS so that the reprojection extent starts at (0, 0)
+            target_wcs.wcs.crpix -= (left, bottom)
+            target_wcs.pixel_shape = (right - left + 1, top - bottom + 1)
+
+            reprojected_map = self.reproject_to(target_wcs)
+
+            # Place the reprojected map correctly in non-shifted pixel coordinates
+            imshow_args['extent'] = (left - 0.5, right + 0.5, bottom - 0.5, top + 0.5)
```

Comment:
Why not pass this in as a kwarg to the imshow call directly?

Reply:
I did it this way because we already do a lot of manipulation of `imshow_args` in this method, and possibly in the future we would refactor the code so that the two imshow code paths (and maybe a third path as well) would simply share the imshow() call.